# Task
Provide a suitable scaler and justification for each of the given features.

## Scaler Recommendations

### Subtask:
Provide a suitable scaler and justification for each of the given features.


## Summary:

The solving process for "Scaler Recommendations" was not provided. Therefore, a summary, key findings, or insights cannot be generated based on the current input.

### Data Analysis Key Findings
* The solving process, which should detail the recommended scalers and their justifications for the given features, was not included in the provided information.

### Insights or Next Steps
* To provide a comprehensive summary and derive meaningful insights, the detailed steps and outcomes of the scaler recommendation process need to be supplied.


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
scores = [20, 25, 30, 50]

# Step 1: Find the minimum and maximum values in the dataset
min_score = min(scores)
max_score = max(scores)

print(f"Original scores: {scores}")
print(f"Minimum score: {min_score}")
print(f"Maximum score: {max_score}")

# Step 2: Apply the Min-Max scaling formula for each score
# Formula: scaled_value = (original_value - min_value) / (max_value - min_value)

scaled_scores = []
print("\nScaling steps:")
for score in scores:
    scaled_value = (score - min_score) / (max_score - min_score)
    scaled_scores.append(scaled_value)
    print(f"  Score {score}: ({score} - {min_score}) / ({max_score} - {min_score}) = {scaled_value:.4f}")

print(f"\nScaled scores: {scaled_scores}")

Original scores: [20, 25, 30, 50]
Minimum score: 20
Maximum score: 50

Scaling steps:
  Score 20: (20 - 20) / (50 - 20) = 0.0000
  Score 25: (25 - 20) / (50 - 20) = 0.1667
  Score 30: (30 - 20) / (50 - 20) = 0.3333
  Score 50: (50 - 20) / (50 - 20) = 1.0000

Scaled scores: [0.0, 0.16666666666666666, 0.3333333333333333, 1.0]


# A3. Z-scores on a subset


In [3]:
x = [8, 9, 11]

print(f"Original data: {x}")

# Step 1: Compute the mean
mean_x = sum(x) / len(x)
print(f"\nMean of x: {mean_x:.2f}")

# Step 2: Compute the population standard deviation
# Calculate the variance first
variance_x = sum([(val - mean_x)**2 for val in x]) / len(x)
std_dev_x = variance_x**0.5
print(f"Population Standard Deviation of x: {std_dev_x:.2f}")

# Step 3: Standardize each value (Z-score)
# Formula: z = (value - mean) / standard_deviation
standardized_x = []
print("\nStandardization steps:")
for val in x:
    z_score = (val - mean_x) / std_dev_x
    standardized_x.append(z_score)
    print(f"  Value {val}: ({val} - {mean_x:.2f}) / {std_dev_x:.2f} = {z_score:.4f}")

print(f"\nStandardized x (Z-scores): {standardized_x}")

Original data: [8, 9, 11]

Mean of x: 9.33
Population Standard Deviation of x: 1.25

Standardization steps:
  Value 8: (8 - 9.33) / 1.25 = -1.0690
  Value 9: (9 - 9.33) / 1.25 = -0.2673
  Value 11: (11 - 9.33) / 1.25 = 1.3363

Standardized x (Z-scores): [-1.069044967649698, -0.2672612419124249, 1.3363062095621216]


## A4. Robust scaling ingredients

In [4]:
import numpy as np

y = [5, 6, 6, 7, 50]
print(f"Original data: {y}")

# Calculate median (Q2)
median_y = np.median(y)

# Calculate Q1 (25th percentile)
Q1_y = np.percentile(y, 25)

# Calculate Q3 (75th percentile)
Q3_y = np.percentile(y, 75)

# Calculate IQR
IQR_y = Q3_y - Q1_y

print(f"Median (Q2): {median_y}")
print(f"Q1: {Q1_y}")
print(f"Q3: {Q3_y}")
print(f"IQR: {IQR_y}")

Original data: [5, 6, 6, 7, 50]
Median (Q2): 6.0
Q1: 6.0
Q3: 7.0
IQR: 1.0


## Outlier Detection using IQR for data `y`

In [5]:
import numpy as np

# Data y and pre-calculated values (from cell rxyBlseNKzeo)
y = [5, 6, 6, 7, 50]
Q1_y = np.percentile(y, 25)
Q3_y = np.percentile(y, 75)
IQR_y = Q3_y - Q1_y

print(f"Original data: {y}")
print(f"Q1: {Q1_y}")
print(f"Q3: {Q3_y}")
print(f"IQR: {IQR_y}")

# Calculate the lower and upper bounds for outliers
lower_bound = Q1_y - 1.5 * IQR_y
upper_bound = Q3_y + 1.5 * IQR_y

print(f"\nLower bound for outliers: {lower_bound}")
print(f"Upper bound for outliers: {upper_bound}")

# Identify outliers
outliers = [val for val in y if val < lower_bound or val > upper_bound]

if outliers:
    print(f"\nIdentified outliers: {outliers}")
else:
    print("\nNo outliers detected.")

Original data: [5, 6, 6, 7, 50]
Q1: 6.0
Q3: 7.0
IQR: 1.0

Lower bound for outliers: 4.5
Upper bound for outliers: 8.5

Identified outliers: [50]


# Part B. Hands on practice
### B1. Three scalers side by side


In [6]:
Heights = [150, 160, 170, 180, 200]
Weights = [50, 60, 70, 80, 200]

from sklearn.preprocessing import MinMaxScaler, StandardScaler, RobustScaler

print("Heights and Weights lists defined, and scalers imported.")

Heights and Weights lists defined, and scalers imported.


## Min-Max Scaling

### Subtask:
Apply Min-Max scaling to both `Heights` and `Weights` to transform their values to the range [0, 1].


In [7]:
from sklearn.preprocessing import MinMaxScaler
import numpy as np

# Reshape data for the scaler
heights_reshaped = np.array(Heights).reshape(-1, 1)
weights_reshaped = np.array(Weights).reshape(-1, 1)

# Initialize MinMaxScaler
minmax_scaler = MinMaxScaler()

# Apply Min-Max scaling to Heights
scaled_heights_minmax = minmax_scaler.fit_transform(heights_reshaped)

# Apply Min-Max scaling to Weights
scaled_weights_minmax = minmax_scaler.fit_transform(weights_reshaped)

print("Original Heights:", Heights)
print("Scaled Heights (Min-Max):")
for original, scaled in zip(Heights, scaled_heights_minmax):
    print(f"  {original} -> {scaled[0]:.4f}")

print("\nOriginal Weights:", Weights)
print("Scaled Weights (Min-Max):")
for original, scaled in zip(Weights, scaled_weights_minmax):
    print(f"  {original} -> {scaled[0]:.4f}")

Original Heights: [150, 160, 170, 180, 200]
Scaled Heights (Min-Max):
  150 -> 0.0000
  160 -> 0.2000
  170 -> 0.4000
  180 -> 0.6000
  200 -> 1.0000

Original Weights: [50, 60, 70, 80, 200]
Scaled Weights (Min-Max):
  50 -> 0.0000
  60 -> 0.0667
  70 -> 0.1333
  80 -> 0.2000
  200 -> 1.0000


## Standard Scaling

### Subtask:
Apply Standard Scaling to `Heights` to transform its values to have a mean of 0 and a standard deviation of 1.



In [8]:
from sklearn.preprocessing import StandardScaler
import numpy as np

# Reshape data for the scaler (Heights)
heights_reshaped = np.array(Heights).reshape(-1, 1)

# Initialize StandardScaler
standard_scaler = StandardScaler()

# Apply Standard Scaling to Heights
scaled_heights_standard = standard_scaler.fit_transform(heights_reshaped)

print("Original Heights:", Heights)
print("Scaled Heights (Standard):")
for original, scaled in zip(Heights, scaled_heights_standard):
    print(f"  {original} -> {scaled[0]:.4f}")

Original Heights: [150, 160, 170, 180, 200]
Scaled Heights (Standard):
  150 -> -1.2787
  160 -> -0.6975
  170 -> -0.1162
  180 -> 0.4650
  200 -> 1.6275


In [9]:
from sklearn.preprocessing import RobustScaler
import numpy as np

# Reshape data for the scaler (Weights)
weights_reshaped = np.array(Weights).reshape(-1, 1)

# Initialize RobustScaler
robust_scaler = RobustScaler()

# Apply Robust Scaling to Weights
scaled_weights_robust = robust_scaler.fit_transform(weights_reshaped)

print("Original Weights:", Weights)
print("Scaled Weights (Robust):")
for original, scaled in zip(Weights, scaled_weights_robust):
    print(f"  {original} -> {scaled[0]:.4f}")

Original Weights: [50, 60, 70, 80, 200]
Scaled Weights (Robust):
  50 -> -1.0000
  60 -> -0.5000
  70 -> 0.0000
  80 -> 0.5000
  200 -> 6.5000


In [10]:
from sklearn.preprocessing import StandardScaler
import numpy as np

# 1. Extract the first three values from Heights and Weights
heights_subset = Heights[:3]
weights_subset = Weights[:3]

print(f"Original Heights subset: {heights_subset}")
print(f"Original Weights subset: {weights_subset}")

# 2. Reshape subsets for StandardScaler
heights_subset_reshaped = np.array(heights_subset).reshape(-1, 1)
weights_subset_reshaped = np.array(weights_subset).reshape(-1, 1)

# 3. Initialize StandardScaler
standard_scaler_subset = StandardScaler()

# 4. Apply Standard Scaling to heights_subset
scaled_heights_subset_standard = standard_scaler_subset.fit_transform(heights_subset_reshaped)

# 5. Apply Standard Scaling to weights_subset
scaled_weights_subset_standard = standard_scaler_subset.fit_transform(weights_subset_reshaped)

print("\nScaled Heights Subset (Standard):")
for original, scaled in zip(heights_subset, scaled_heights_subset_standard):
    print(f"  {original} -> {scaled[0]:.4f}")

print("\nScaled Weights Subset (Standard):")
for original, scaled in zip(weights_subset, scaled_weights_subset_standard):
    print(f"  {original} -> {scaled[0]:.4f}")

Original Heights subset: [150, 160, 170]
Original Weights subset: [50, 60, 70]

Scaled Heights Subset (Standard):
  150 -> -1.2247
  160 -> 0.0000
  170 -> 1.2247

Scaled Weights Subset (Standard):
  50 -> -1.2247
  60 -> 0.0000
  70 -> 1.2247


## B2. One-hot by hand

In [11]:
import pandas as pd

Cities = ['Dhaka', 'Chattogram', 'Dhaka', 'Rajshahi', 'Rajshahi']

# Get unique cities to create column names
unique_cities = sorted(list(set(Cities)))

# Create a dictionary to store the one-hot encoded data
one_hot_data = {}
for city_name in unique_cities:
    one_hot_data[f'City_{city_name}'] = []

# Populate the one-hot encoded data
for city in Cities:
    for city_name in unique_cities:
        if city == city_name:
            one_hot_data[f'City_{city_name}'].append(1)
        else:
            one_hot_data[f'City_{city_name}'].append(0)

# Create a DataFrame
one_hot_df = pd.DataFrame(one_hot_data)

print("Original Cities:", Cities)
print("\nOne-hot encoded DataFrame:")
print(one_hot_df)

Original Cities: ['Dhaka', 'Chattogram', 'Dhaka', 'Rajshahi', 'Rajshahi']

One-hot encoded DataFrame:
   City_Chattogram  City_Dhaka  City_Rajshahi
0                0           1              0
1                1           0              0
2                0           1              0
3                0           0              1
4                0           0              1


## B3. Ordinal mapping

In [12]:
Education = ['High School', 'Bachelor', 'Master', 'Bachelor', 'Master']

# Mapping 1: High School=0, Bachelor=1, Master=2
mapping_1 = {'High School': 0, 'Bachelor': 1, 'Master': 2}
encoded_education_1 = [mapping_1[edu] for edu in Education]
print("Original Education:", Education)
print("Encoded Education (Map 0,1,2):", encoded_education_1)

# Mapping 2: High School=1, Bachelor=2, Master=3
mapping_2 = {'High School': 1, 'Bachelor': 2, 'Master': 3}
encoded_education_2 = [mapping_2[edu] for edu in Education]
print("\nEncoded Education (Map 1,2,3):", encoded_education_2)

# Explanation of distance shift:
# Shifting the map by adding a constant to each category value (e.g., from 0,1,2 to 1,2,3) does not change the *relative* distances between any two points, as the absolute difference remains the same.

Original Education: ['High School', 'Bachelor', 'Master', 'Bachelor', 'Master']
Encoded Education (Map 0,1,2): [0, 1, 2, 1, 2]

Encoded Education (Map 1,2,3): [1, 2, 3, 2, 3]


## B5. Vectors and alignment

In [13]:
import numpy as np

# Define the vectors
a = np.array([3, -1, 2])
b = np.array([4, 0, -2])
c = np.array([-6, 2, -4])

print(f"Vector a: {a}")
print(f"Vector b: {b}")
print(f"Vector c: {c}\n")

# a) Compute a·b and a·c
dot_ab = np.dot(a, b)
dot_ac = np.dot(a, c)

print(f"a·b = {dot_ab}")
print(f"a·c = {dot_ac}\n")

# b) Compare signs and magnitudes to comment on the alignment
print("Alignment Analysis:")

# Alignment of a with b
if dot_ab > 0:
    print(f"- a and b have a positive dot product ({dot_ab}), indicating they are generally aligned (angle < 90 degrees).")
elif dot_ab < 0:
    print(f"- a and b have a negative dot product ({dot_ab}), indicating they are generally anti-aligned (angle > 90 degrees).")
else:
    print(f"- a and b have a dot product of zero, indicating they are orthogonal (perpendicular).")

# Alignment of a with c
if dot_ac > 0:
    print(f"- a and c have a positive dot product ({dot_ac}), indicating they are generally aligned (angle < 90 degrees).")
elif dot_ac < 0:
    print(f"- a and c have a negative dot product ({dot_ac}), indicating they are generally anti-aligned (angle > 90 degrees).")
else:
    print(f"- a and c have a dot product of zero, indicating they are orthogonal (perpendicular).")

print(f"  Since a·c ({dot_ac}) is exactly -2 times a·b ({dot_ab}), it suggests c is exactly anti-parallel to a with a proportional magnitude relative to b.\n")

# c) L2 normalize a and give the normalized vector to three decimals
# Calculate the L2 norm (Euclidean norm)
norm_a = np.linalg.norm(a)
normalized_a = a / norm_a

print(f"L2 Norm of a: {norm_a:.4f}")
print(f"Normalized vector a (to three decimals): {np.round(normalized_a, 3)}")

Vector a: [ 3 -1  2]
Vector b: [ 4  0 -2]
Vector c: [-6  2 -4]

a·b = 8
a·c = -28

Alignment Analysis:
- a and b have a positive dot product (8), indicating they are generally aligned (angle < 90 degrees).
- a and c have a negative dot product (-28), indicating they are generally anti-aligned (angle > 90 degrees).
  Since a·c (-28) is exactly -2 times a·b (8), it suggests c is exactly anti-parallel to a with a proportional magnitude relative to b.

L2 Norm of a: 3.7417
Normalized vector a (to three decimals): [ 0.802 -0.267  0.535]


## B6. Two distances, different vibes

In [14]:
import numpy as np
from scipy.spatial.distance import euclidean, cityblock # cityblock is Manhattan distance

# Define the points
P1 = np.array([2, 3])
P2 = np.array([5, 7])
P3 = np.array([2, 10])

print(f"P1: {P1}, P2: {P2}, P3: {P3}\n")

# a) Compute Euclidean and Manhattan distances for all pairs

# P1 and P2
d_euclidean_p1p2 = euclidean(P1, P2)
d_manhattan_p1p2 = cityblock(P1, P2)
print(f"Distances between P1 and P2:")
print(f"  Euclidean: {d_euclidean_p1p2:.4f}")
print(f"  Manhattan: {d_manhattan_p1p2:.4f}\n")

# P1 and P3
d_euclidean_p1p3 = euclidean(P1, P3)
d_manhattan_p1p3 = cityblock(P1, P3)
print(f"Distances between P1 and P3:")
print(f"  Euclidean: {d_euclidean_p1p3:.4f}")
print(f"  Manhattan: {d_manhattan_p1p3:.4f}\n")

# P2 and P3
d_euclidean_p2p3 = euclidean(P2, P3)
d_manhattan_p2p3 = cityblock(P2, P3)
print(f"Distances between P2 and P3:")
print(f"  Euclidean: {d_euclidean_p2p3:.4f}")
print(f"  Manhattan: {d_manhattan_p2p3:.4f}\n")

# b) Which distance is more sensitive to a single large jump in one coordinate
print("b) Sensitivity to a single large jump:")
print("   Manhattan distance is more sensitive to a single large jump in one coordinate because it sums the absolute differences across dimensions, while Euclidean distance squares these differences, diminishing the relative impact of a single large jump compared to multiple smaller changes.")

# c) Scale y by 10 and recompute d(P1, P2) for both distances, then explain the effect in one line

P1_scaled = np.array([P1[0], P1[1] * 10])
P2_scaled = np.array([P2[0], P2[1] * 10])

print(f"\nScaled P1: {P1_scaled}, Scaled P2: {P2_scaled}\n")

d_euclidean_p1p2_scaled = euclidean(P1_scaled, P2_scaled)
d_manhattan_p1p2_scaled = cityblock(P1_scaled, P2_scaled)

print(f"Distances between Scaled P1 and Scaled P2:")
print(f"  Euclidean: {d_euclidean_p1p2_scaled:.4f}")
print(f"  Manhattan: {d_manhattan_p1p2_scaled:.4f}\n")

print("c) Effect of scaling on distances:")
print("   Scaling a coordinate by 10 increases both Euclidean and Manhattan distances by a factor that depends on the magnitude of the scaled coordinate's difference, effectively amplifying its contribution to the overall distance.")


P1: [2 3], P2: [5 7], P3: [ 2 10]

Distances between P1 and P2:
  Euclidean: 5.0000
  Manhattan: 7.0000

Distances between P1 and P3:
  Euclidean: 7.0000
  Manhattan: 7.0000

Distances between P2 and P3:
  Euclidean: 4.2426
  Manhattan: 6.0000

b) Sensitivity to a single large jump:
   Manhattan distance is more sensitive to a single large jump in one coordinate because it sums the absolute differences across dimensions, while Euclidean distance squares these differences, diminishing the relative impact of a single large jump compared to multiple smaller changes.

Scaled P1: [ 2 30], Scaled P2: [ 5 70]

Distances between Scaled P1 and Scaled P2:
  Euclidean: 40.1123
  Manhattan: 43.0000

c) Effect of scaling on distances:
   Scaling a coordinate by 10 increases both Euclidean and Manhattan distances by a factor that depends on the magnitude of the scaled coordinate's difference, effectively amplifying its contribution to the overall distance.


# Part C. Mini datasets

# Task
```python
C_Data_1 = pd.DataFrame({
    'ID': [1, 2, 3, 4, 5],
    'Age': [25, 30, 22, 28, 35],
    'Education': ['Bachelor', 'Master', 'High School', 'Bachelor', 'PhD'],
    'Internet': ['DSL', 'Fiber', 'Dial-up', 'Fiber', 'DSL'],
    'Hours_Study': [30, 40, 20, 35, 45],
    'GPA': [3.5, 3.8, 3.0, 3.6, 3.9],
    'Satisfaction': ['High', 'Very High', 'Medium', 'High', 'Very High']
})

print("C-Data-1 DataFrame created:")
print(C_Data_1)
```

## Create C-Data-1 DataFrame

### Subtask:
Create a pandas DataFrame for C-Data-1 as provided, using the specified columns and data.


**Reasoning**:
The subtask is to create a pandas DataFrame named `C_Data_1` with the specified columns and data. I will use the `code_block` command to create this DataFrame and then print it to verify its content.



In [15]:
import pandas as pd

# Create the C_Data_1 DataFrame
C_Data_1 = pd.DataFrame({
    'ID': [1, 2, 3, 4, 5],
    'Age': [25, 30, 22, 28, 35],
    'Education': ['Bachelor', 'Master', 'High School', 'Bachelor', 'PhD'],
    'Internet': ['DSL', 'Fiber', 'Dial-up', 'Fiber', 'DSL'],
    'Hours_Study': [30, 40, 20, 35, 45],
    'GPA': [3.5, 3.8, 3.0, 3.6, 3.9],
    'Satisfaction': ['High', 'Very High', 'Medium', 'High', 'Very High']
})

# Print the DataFrame to verify its content
print(C_Data_1)

   ID  Age    Education Internet  Hours_Study  GPA Satisfaction
0   1   25     Bachelor      DSL           30  3.5         High
1   2   30       Master    Fiber           40  3.8    Very High
2   3   22  High School  Dial-up           20  3.0       Medium
3   4   28     Bachelor    Fiber           35  3.6         High
4   5   35          PhD      DSL           45  3.9    Very High


## Create C-Data-2 DataFrame

### Subtask:
Create a pandas DataFrame for C-Data-2 as provided, using the specified columns and data.


## Create C-Data-2 DataFrame

### Subtask:
Create a pandas DataFrame for C-Data-2 as provided, using the specified columns and data.


**Reasoning**:
The subtask is to create a pandas DataFrame named `C_Data_2` with the specified columns and data. I will use the `code_block` command to create this DataFrame and then print it to verify its content.



In [16]:
import pandas as pd

# Create the C_Data_2 DataFrame
C_Data_2 = pd.DataFrame({
    'ID': [1, 2, 3, 4, 5],
    'Income_BDT': [50000, 60000, 45000, 70000, 300000],
    'Transactions': [10, 12, 8, 15, 50],
    'City': ['Dhaka', 'Chattogram', 'Dhaka', 'Rajshahi', 'Khulna'],
    'Temp_C': [28, 30, 27, 32, 29],
    'Rating': [4, 5, 3, 4, 5]
})

# Print the DataFrame to verify its content
print(C_Data_2)

   ID  Income_BDT  Transactions        City  Temp_C  Rating
0   1       50000            10       Dhaka      28       4
1   2       60000            12  Chattogram      30       5
2   3       45000             8       Dhaka      27       3
3   4       70000            15    Rajshahi      32       4
4   5      300000            50      Khulna      29       5


## C1. Scaler choices for Income_BDT, Transactions, Temp_C

### Subtask:
For each of the specified features from C-Data-2 (Income_BDT, Transactions, Temp_C), select a suitable scaler, provide a one-line justification for the choice, and include a two-line numeric illustration using the feature's values. RobustScaler will be chosen for Income_BDT and Transactions due to outliers, and StandardScaler for Temp_C.


### Income_BDT: RobustScaler

**Justification**: `RobustScaler` is chosen for 'Income_BDT' because the data visually contains a significant outlier (300,000 compared to others in the 45,000-70,000 range), and `RobustScaler` handles outliers effectively by using the median and Interquartile Range (IQR).

**Reasoning**:
I will now initialize and fit the `RobustScaler` to the 'Income_BDT' column, print the original values, and then illustrate the scaling with two specific values as requested in the subtask.



In [17]:
from sklearn.preprocessing import RobustScaler
import numpy as np

# Extract 'Income_BDT' column and reshape for scaler
income_bdt = C_Data_2['Income_BDT'].values.reshape(-1, 1)

# Initialize RobustScaler
robust_scaler_income = RobustScaler()

# Apply Robust Scaling to Income_BDT
scaled_income_bdt = robust_scaler_income.fit_transform(income_bdt)

print("Original Income_BDT values:", C_Data_2['Income_BDT'].tolist())
print("Scaled Income_BDT values (RobustScaler):")

# Numeric illustration using the first and last values
original_val_1 = C_Data_2['Income_BDT'].iloc[0]
scaled_val_1 = scaled_income_bdt[0][0]
original_val_2 = C_Data_2['Income_BDT'].iloc[-1]
scaled_val_2 = scaled_income_bdt[-1][0]

print(f"  {original_val_1} -> {scaled_val_1:.4f}")
print(f"  {original_val_2} -> {scaled_val_2:.4f}")

Original Income_BDT values: [50000, 60000, 45000, 70000, 300000]
Scaled Income_BDT values (RobustScaler):
  50000 -> -0.5000
  300000 -> 12.0000


### Transactions: RobustScaler

**Justification**: `RobustScaler` is chosen for 'Transactions' because the data contains a clear outlier (50 compared to others in the 8-15 range), and `RobustScaler` effectively handles outliers by using the median and Interquartile Range (IQR).

**Reasoning**:
Now I will apply Robust Scaling to the 'Transactions' column, print the original values, and illustrate the scaling with two specific values from the data, as outlined in the subtask instructions.



In [18]:
from sklearn.preprocessing import RobustScaler
import numpy as np

# Extract 'Transactions' column and reshape for scaler
transactions = C_Data_2['Transactions'].values.reshape(-1, 1)

# Initialize RobustScaler
robust_scaler_transactions = RobustScaler()

# Apply Robust Scaling to Transactions
scaled_transactions = robust_scaler_transactions.fit_transform(transactions)

print("Original Transactions values:", C_Data_2['Transactions'].tolist())
print("Scaled Transactions values (RobustScaler):")

# Numeric illustration using the first and last values
original_val_1 = C_Data_2['Transactions'].iloc[0]
scaled_val_1 = scaled_transactions[0][0]
original_val_2 = C_Data_2['Transactions'].iloc[-1]
scaled_val_2 = scaled_transactions[-1][0]

print(f"  {original_val_1} -> {scaled_val_1:.4f}")
print(f"  {original_val_2} -> {scaled_val_2:.4f}")

Original Transactions values: [10, 12, 8, 15, 50]
Scaled Transactions values (RobustScaler):
  10 -> -0.4000
  50 -> 7.6000


### Temp_C: StandardScaler

**Justification**: `StandardScaler` is chosen for 'Temp_C' because the data appears to have no significant outliers and a relatively normal distribution, making it suitable for transformation to a mean of 0 and standard deviation of 1.

### Temp_C: StandardScaler

**Justification**: `StandardScaler` is chosen for 'Temp_C' because the data appears to have no significant outliers and a relatively normal distribution, making it suitable for transformation to a mean of 0 and standard deviation of 1.

**Reasoning**:
Now, I will apply Standard Scaling to the 'Temp_C' column, print the original values, and illustrate the scaling with two specific values from the data, as outlined in the subtask instructions.



In [19]:
from sklearn.preprocessing import StandardScaler
import numpy as np

# Extract 'Temp_C' column and reshape for scaler
temp_c = C_Data_2['Temp_C'].values.reshape(-1, 1)

# Initialize StandardScaler
standard_scaler_temp_c = StandardScaler()

# Apply Standard Scaling to Temp_C
scaled_temp_c = standard_scaler_temp_c.fit_transform(temp_c)

print("Original Temp_C values:", C_Data_2['Temp_C'].tolist())
print("Scaled Temp_C values (StandardScaler):")

# Numeric illustration using the first and last values
original_val_1 = C_Data_2['Temp_C'].iloc[0]
scaled_val_1 = scaled_temp_c[0][0]
original_val_2 = C_Data_2['Temp_C'].iloc[-1]
scaled_val_2 = scaled_temp_c[-1][0]

print(f"  {original_val_1} -> {scaled_val_1:.4f}")
print(f"  {original_val_2} -> {scaled_val_2:.4f}")

Original Temp_C values: [28, 30, 27, 32, 29]
Scaled Temp_C values (StandardScaler):
  28 -> -0.6975
  29 -> -0.1162


## C2. Mixed preprocessing plan

### Subtask:
For C-Data-1 and C-Data-2 combined: identify nominal and ordinal columns; propose an encoding plan listing columns for one-hot encoding (Internet, City) vs. ordinal encoding (Education, Satisfaction with mappings); and propose a scaling plan listing columns for Min-Max (Age, GPA), Standardization (Hours_Study, Temp_C), and Robust scaling (Income_BDT, Transactions).


### Nominal Categorical Columns (One-Hot Encoding)

**Columns**: `Internet` (from C-Data-1) and `City` (from C-Data-2)

**Reasoning**: These columns represent distinct categories without any inherent order. One-hot encoding is suitable for converting these nominal categories into a numerical format that machine learning algorithms can understand, preventing the model from assuming an arbitrary ordinal relationship.

**Encoding Plan**: Apply one-hot encoding to `Internet` and `City`.

### Ordinal Categorical Columns (Ordinal Encoding)

**Columns**: `Education` (from C-Data-1), `Satisfaction` (from C-Data-1), and `Rating` (from C-Data-2)

**Reasoning**: These columns represent categories with a clear, intrinsic order. Ordinal encoding preserves this order by mapping categories to numerical values, which is appropriate for machine learning models.

**Encoding Plan**:
*   `Education`: Map as {'High School': 0, 'Bachelor': 1, 'Master': 2, 'PhD': 3}.
*   `Satisfaction`: Map as {'Medium': 0, 'High': 1, 'Very High': 2}.
*   `Rating`: Map as {3: 0, 4: 1, 5: 2}.

### Numerical Columns for Min-Max Scaling

**Columns**: `Age` (from C-Data-1) and `GPA` (from C-Data-1)

**Reasoning**: These columns are numerical, and based on the provided data, they do not appear to have significant outliers. Min-Max scaling is appropriate here to transform the values into a fixed range (e.g., 0 to 1), which is useful for algorithms sensitive to feature scales, such as neural networks.

**Scaling Plan**: Apply Min-Max scaling to `Age` and `GPA`.

### Numerical Columns for Standard Scaling

**Columns**: `Hours_Study` (from C-Data-1) and `Temp_C` (from C-Data-2)

**Reasoning**: These columns are numerical and appear to have a relatively normal distribution without significant outliers. StandardScaler is suitable for transforming these values to have a mean of 0 and a standard deviation of 1, which is beneficial for many machine learning algorithms that assume normally distributed data or are sensitive to feature variances.

**Scaling Plan**: Apply StandardScaler to `Hours_Study` and `Temp_C`.

### Numerical Columns for Robust Scaling

**Columns**: `Income_BDT` (from C-Data-2) and `Transactions` (from C-Data-2)

**Reasoning**: These columns are numerical and contain clear outliers (e.g., 'Income_BDT' has 300,000 while others are around 50,000-70,000; 'Transactions' has 50 while others are around 8-15). RobustScaler is ideal for handling such data as it scales features using the median and Interquartile Range (IQR), making it robust to outliers.

**Scaling Plan**: Apply RobustScaler to `Income_BDT` and `Transactions`.

### Numerical Columns for Robust Scaling

**Columns**: `Income_BDT` (from C-Data-2) and `Transactions` (from C-Data-2)

**Reasoning**: These columns are numerical and contain clear outliers (e.g., 'Income_BDT' has 300,000 while others are around 50,000-70,000; 'Transactions' has 50 while others are around 8-15). RobustScaler is ideal for handling such data as it scales features using the median and Interquartile Range (IQR), making it robust to outliers.

**Scaling Plan**: Apply RobustScaler to `Income_BDT` and `Transactions`.

## C3. Outlier stress test for Income_BDT

### Subtask:
Using Income_BDT from C-Data-2, compute Min-Max scaled values and Robust scaled values. Then, in one line, compare how each scaler treats the outlier value (300000).


**Reasoning**:
I will extract the 'Income_BDT' column, reshape it, then apply both `MinMaxScaler` and `RobustScaler` to it. After scaling, I will print the original and scaled values for both methods to prepare for the comparison of how they treat the outlier, as requested by the subtask.



In [20]:
from sklearn.preprocessing import MinMaxScaler, RobustScaler
import numpy as np

# Extract 'Income_BDT' column and reshape for scaler
income_bdt_data = C_Data_2['Income_BDT'].values.reshape(-1, 1)

# Initialize and apply MinMaxScaler
minmax_scaler_income = MinMaxScaler()
minmax_scaled_income_bdt = minmax_scaler_income.fit_transform(income_bdt_data)

# Initialize and apply RobustScaler
robust_scaler_income = RobustScaler()
robust_scaled_income_bdt = robust_scaler_income.fit_transform(income_bdt_data)

print("Original Income_BDT values:", C_Data_2['Income_BDT'].tolist())

print("\nMin-Max Scaled Income_BDT values:")
for original, scaled in zip(C_Data_2['Income_BDT'], minmax_scaled_income_bdt):
    print(f"  {original} -> {scaled[0]:.4f}")

print("\nRobust Scaled Income_BDT values:")
for original, scaled in zip(C_Data_2['Income_BDT'], robust_scaled_income_bdt):
    print(f"  {original} -> {scaled[0]:.4f}")

Original Income_BDT values: [50000, 60000, 45000, 70000, 300000]

Min-Max Scaled Income_BDT values:
  50000 -> 0.0196
  60000 -> 0.0588
  45000 -> 0.0000
  70000 -> 0.0980
  300000 -> 1.0000

Robust Scaled Income_BDT values:
  50000 -> -0.5000
  60000 -> 0.0000
  45000 -> -0.7500
  70000 -> 0.5000
  300000 -> 12.0000


Comparing how scalers treat the outlier (300000):
Min-Max scaling compresses the outlier (300000) to the maximum value of 1.0, while Robust scaling allows the outlier to significantly exceed the typical range, resulting in a value of 12.0, thereby preserving its outlying nature.

## C4. Distance on feature space for Hours_Study and GPA

### Subtask:
From C-Data-1, extract 'Hours_Study' and 'GPA' for ID 1 and ID 4. Compute the Euclidean distance and Manhattan distance for this pair. Then, normalize both features using Min-Max scaling, recompute both distances, and comment on the effect of scaling on the distances in one line.


## Summary:

### Data Analysis Key Findings

*   **DataFrame Creation:** Both `C_Data_1` and `C_Data_2` DataFrames were successfully initialized and verified. The creation of `C_Data_2` required a re-attempt after initial missing specifications for its content.
*   **Feature Scaling Selection (C1):**
    *   `RobustScaler` was appropriately chosen for 'Income\_BDT' and 'Transactions' due to the presence of significant outliers. For 'Income\_BDT', the outlier value 300,000 was scaled to 12.0000, and for 'Transactions', the outlier 50 scaled to 7.6000, demonstrating its robustness.
    *   `StandardScaler` was selected for 'Temp\_C' (e.g., 28 -> -0.6975), justified by its apparent normal distribution and lack of significant outliers.
*   **Mixed Preprocessing Plan (C2):** A comprehensive plan was developed for combining `C_Data_1` and `C_Data_2` features:
    *   **Nominal Features:** 'Internet' (from `C_Data_1`) and 'City' (from `C_Data_2`) were designated for One-Hot Encoding.
    *   **Ordinal Features:** 'Education' (from `C_Data_1`), 'Satisfaction' (from `C_Data_1`), and 'Rating' (from `C_Data_2`) were assigned Ordinal Encoding with specific value mappings (e.g., 'High School': 0 for Education; 'Medium': 0 for Satisfaction; 3: 0 for Rating).
    *   **Numerical Scaling:**
        *   `Min-Max Scaling` for 'Age' and 'GPA' (from `C_Data_1`) for features without significant outliers.
        *   `Standard Scaling` for 'Hours\_Study' (from `C_Data_1`) and 'Temp\_C' (from `C_Data_2`) for normally distributed data.
        *   `Robust Scaling` for 'Income\_BDT' and 'Transactions' (from `C_Data_2`) to effectively manage outliers.
*   **Outlier Stress Test for Income\_BDT (C3):**
    *   `MinMaxScaler` treated the 300,000 outlier in 'Income\_BDT' by compressing it to the maximum scaled value of 1.0000, effectively reducing its unique impact on the scale.
    *   Conversely, `RobustScaler` transformed the same outlier to 12.0000, which significantly preserves its outlying nature relative to other data points, highlighting its differential treatment compared to `MinMaxScaler`.

### Insights or Next Steps

*   The careful selection of preprocessing techniques, including robust scaling for outlier-prone features, is critical for preparing diverse datasets for machine learning models, ensuring that data characteristics are appropriately handled.
*   Further analysis should proceed with implementing the proposed encoding and scaling plans to transform the raw data, followed by feature engineering or model training, to evaluate the practical impact of these preprocessing choices.


# Task
Apply `ColumnTransformer` for mixed preprocessing: Combine `C_Data_1` and `C_Data_2` DataFrames, then apply one-hot encoding to nominal features (`Internet`, `City`), ordinal encoding to ordinal features (`Education` with {'High School': 0, 'Bachelor': 1, 'Master': 2, 'PhD': 3}, `Satisfaction` with {'Medium': 0, 'High': 1, 'Very High': 2}, `Rating` with {3: 0, 4: 1, 5: 2}), Min-Max scaling to `Age` and `GPA`, Standard scaling to `Hours_Study` and `Temp_C`, and Robust scaling to `Income_BDT` and `Transactions`. Display the shape and first five rows of the resulting transformed data.

## Apply ColumnTransformer for mixed preprocessing

### Subtask:
Combine C-Data-1 and C-Data-2, then use sklearn's ColumnTransformer to apply one-hot encoding to nominal features, ordinal encoding to ordinal features (with specified mappings), and appropriate scaling (Min-Max, Standard, Robust) to numerical features. Display the shape and first few rows of the transformed data.


**Reasoning**:
The first step is to import all the necessary libraries as specified in the instructions to prepare for data merging and transformation.



In [21]:
import pandas as pd
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder, MinMaxScaler, StandardScaler, RobustScaler

print("Required libraries imported successfully.")

Required libraries imported successfully.


**Reasoning**:
Now that the libraries are imported, I will merge C_Data_1 and C_Data_2 into a single DataFrame `combined_df` using the 'ID' column, and then drop the 'ID' column as it's not a feature for transformation.



In [22]:
combined_df = pd.merge(C_Data_1, C_Data_2, on='ID', how='inner')
combined_df = combined_df.drop('ID', axis=1)

print("Combined DataFrame created and 'ID' column dropped. Shape:", combined_df.shape)
print("First 5 rows of combined_df:\n", combined_df.head())

Combined DataFrame created and 'ID' column dropped. Shape: (5, 11)
First 5 rows of combined_df:
    Age    Education Internet  Hours_Study  GPA Satisfaction  Income_BDT  \
0   25     Bachelor      DSL           30  3.5         High       50000   
1   30       Master    Fiber           40  3.8    Very High       60000   
2   22  High School  Dial-up           20  3.0       Medium       45000   
3   28     Bachelor    Fiber           35  3.6         High       70000   
4   35          PhD      DSL           45  3.9    Very High      300000   

   Transactions        City  Temp_C  Rating  
0            10       Dhaka      28       4  
1            12  Chattogram      30       5  
2             8       Dhaka      27       3  
3            15    Rajshahi      32       4  
4            50      Khulna      29       5  


**Reasoning**:
Now that the `combined_df` is prepared, I will define the column lists for nominal, ordinal, Min-Max, Standard, and Robust scaling, along with the specific category orderings for ordinal features, as outlined in the subtask instructions.



In [23]:
nominal_features = ['Internet', 'City']
ordinal_features = ['Education', 'Satisfaction', 'Rating']
minmax_features = ['Age', 'GPA']
standard_features = ['Hours_Study', 'Temp_C']
robust_features = ['Income_BDT', 'Transactions']

education_categories = ['High School', 'Bachelor', 'Master', 'PhD']
satisfaction_categories = ['Medium', 'High', 'Very High']
rating_categories = [3, 4, 5]

print("Feature lists and ordinal categories defined.")

Feature lists and ordinal categories defined.


**Reasoning**:
With the feature lists and ordinal categories defined, I will now create the `ColumnTransformer` to apply all specified preprocessing steps (one-hot encoding, ordinal encoding, Min-Max scaling, Standard scaling, and Robust scaling) to the `combined_df`. Then, I will fit and transform the data and display the shape and first five rows of the resulting transformed data as per the instructions.



In [24]:
preprocessor = ColumnTransformer(
    transformers=[
        ('onehot', OneHotEncoder(handle_unknown='ignore'), nominal_features),
        ('ordinal', OrdinalEncoder(categories=[education_categories, satisfaction_categories, rating_categories]), ordinal_features),
        ('minmax', MinMaxScaler(), minmax_features),
        ('standard', StandardScaler(), standard_features),
        ('robust', RobustScaler(), robust_features)
    ])

transformed_data = preprocessor.fit_transform(combined_df)

print("Shape of transformed_data:", transformed_data.shape)
print("\nFirst 5 rows of transformed_data:\n", transformed_data[:5])

Shape of transformed_data: (5, 16)

First 5 rows of transformed_data:
 [[ 1.          0.          0.          0.          1.          0.
   0.          1.          1.          1.          0.23076923  0.55555556
  -0.46499055 -0.69748583 -0.5        -0.4       ]
 [ 0.          0.          1.          1.          0.          0.
   0.          2.          2.          2.          0.61538462  0.88888889
   0.69748583  0.46499055  0.          0.        ]
 [ 0.          1.          0.          0.          1.          0.
   0.          0.          0.          0.          0.          0.
  -1.62746694 -1.27872403 -0.75       -0.8       ]
 [ 0.          0.          1.          0.          0.          0.
   1.          1.          1.          1.          0.46153846  0.66666667
   0.11624764  1.62746694  0.5         0.6       ]
 [ 1.          0.          0.          0.          0.          1.
   0.          3.          2.          2.          1.          1.
   1.27872403 -0.11624764 12.          7.

## Distance analysis before and after scaling

### Subtask:
Select 'Income_BDT' and 'Transactions' from C-Data-2 for the first three rows (P1, P2, P3). Compute Euclidean and Manhattan distances between all pairs before scaling. Then, apply StandardScaler and RobustScaler separately to these two features, recompute both distances for all pairs after each scaling, and present the results in a markdown table.


**Reasoning**:
I will start by extracting the specified columns ('Income_BDT' and 'Transactions') for the first three rows from `C_Data_2` to create the `subset_data` as instructed.



In [25]:
import numpy as np
import pandas as pd

# Extract 'Income_BDT' and 'Transactions' for the first three rows
subset_data = C_Data_2[['Income_BDT', 'Transactions']].iloc[:3].values

print("Extracted subset data (first three rows of Income_BDT and Transactions):")
print(subset_data)

Extracted subset data (first three rows of Income_BDT and Transactions):
[[50000    10]
 [60000    12]
 [45000     8]]


**Reasoning**:
I will import the necessary distance metrics and then compute the Euclidean and Manhattan distances for all pairs of points in the `subset_data` as instructed, storing these results for the final table.



In [26]:
from scipy.spatial.distance import euclidean, cityblock

# Define P1, P2, P3 from subset_data
P1_unscaled = subset_data[0]
P2_unscaled = subset_data[1]
P3_unscaled = subset_data[2]

# Calculate distances between all pairs for unscaled data
d_euclidean_P1P2_unscaled = euclidean(P1_unscaled, P2_unscaled)
d_manhattan_P1P2_unscaled = cityblock(P1_unscaled, P2_unscaled)

d_euclidean_P1P3_unscaled = euclidean(P1_unscaled, P3_unscaled)
d_manhattan_P1P3_unscaled = cityblock(P1_unscaled, P3_unscaled)

d_euclidean_P2P3_unscaled = euclidean(P2_unscaled, P3_unscaled)
d_manhattan_P2P3_unscaled = cityblock(P2_unscaled, P3_unscaled)

print("Distances calculated for unscaled data and stored.")

Distances calculated for unscaled data and stored.


**Reasoning**:
Now I will initialize and apply `StandardScaler` to the `subset_data`, then recompute the Euclidean and Manhattan distances for all pairs of points using this standard-scaled data, as required by the subtask instructions.



In [27]:
from sklearn.preprocessing import StandardScaler

# Initialize StandardScaler
standard_scaler = StandardScaler()

# Apply Standard Scaling to subset_data
scaled_data_standard = standard_scaler.fit_transform(subset_data)

# Define P1, P2, P3 from standard-scaled data
P1_standard_scaled = scaled_data_standard[0]
P2_standard_scaled = scaled_data_standard[1]
P3_standard_scaled = scaled_data_standard[2]

# Calculate distances between all pairs for standard-scaled data
d_euclidean_P1P2_standard = euclidean(P1_standard_scaled, P2_standard_scaled)
d_manhattan_P1P2_standard = cityblock(P1_standard_scaled, P2_standard_scaled)

d_euclidean_P1P3_standard = euclidean(P1_standard_scaled, P3_standard_scaled)
d_manhattan_P1P3_standard = cityblock(P1_standard_scaled, P3_standard_scaled)

d_euclidean_P2P3_standard = euclidean(P2_standard_scaled, P3_standard_scaled)
d_manhattan_P2P3_standard = cityblock(P2_standard_scaled, P3_standard_scaled)

print("Distances calculated for standard-scaled data and stored.")

Distances calculated for standard-scaled data and stored.


**Reasoning**:
Now, I will initialize and apply `RobustScaler` to the original `subset_data`, and then recompute the Euclidean and Manhattan distances for all pairs of points using this robust-scaled data, as required by the subtask instructions.



In [28]:
from sklearn.preprocessing import RobustScaler

# Initialize RobustScaler
robust_scaler = RobustScaler()

# Apply Robust Scaling to subset_data
scaled_data_robust = robust_scaler.fit_transform(subset_data)

# Define P1, P2, P3 from robust-scaled data
P1_robust_scaled = scaled_data_robust[0]
P2_robust_scaled = scaled_data_robust[1]
P3_robust_scaled = scaled_data_robust[2]

# Calculate distances between all pairs for robust-scaled data
d_euclidean_P1P2_robust = euclidean(P1_robust_scaled, P2_robust_scaled)
d_manhattan_P1P2_robust = cityblock(P1_robust_scaled, P2_robust_scaled)

d_euclidean_P1P3_robust = euclidean(P1_robust_scaled, P3_robust_scaled)
d_manhattan_P1P3_robust = cityblock(P1_robust_scaled, P3_robust_scaled)

d_euclidean_P2P3_robust = euclidean(P2_robust_scaled, P3_robust_scaled)
d_manhattan_P2P3_robust = cityblock(P2_robust_scaled, P3_robust_scaled)

print("Distances calculated for robust-scaled data and stored.")

Distances calculated for robust-scaled data and stored.


### Distance Comparison for 'Income_BDT' and 'Transactions' (first 3 rows)

| Pair       | Euclidean (Unscaled) | Manhattan (Unscaled) | Euclidean (Standard) | Manhattan (Standard) | Euclidean (Robust) | Manhattan (Robust) |
|:-----------|:---------------------|:---------------------|:---------------------|:---------------------|:-------------------|:-------------------|
| P1 vs P2   | {d_euclidean_P1P2_unscaled:.4f}     | {d_manhattan_P1P2_unscaled:.4f}     | {d_euclidean_P1P2_standard:.4f}     | {d_manhattan_P1P2_standard:.4f}     | {d_euclidean_P1P2_robust:.4f}   | {d_manhattan_P1P2_robust:.4f}   |
| P1 vs P3   | {d_euclidean_P1P3_unscaled:.4f}     | {d_manhattan_P1P3_unscaled:.4f}     | {d_euclidean_P1P3_standard:.4f}     | {d_manhattan_P1P3_standard:.4f}     | {d_euclidean_P1P3_robust:.4f}   | {d_manhattan_P1P3_robust:.4f}   |
| P2 vs P3   | {d_euclidean_P2P3_unscaled:.4f}     | {d_manhattan_P2P3_unscaled:.4f}     | {d_euclidean_P2P3_standard:.4f}     | {d_manhattan_P2P3_standard:.4f}     | {d_euclidean_P2P3_robust:.4f}   | {d_manhattan_P2P3_robust:.4f}   |


### Distance Comparison for 'Income_BDT' and 'Transactions' (first 3 rows)

| Pair       | Euclidean (Unscaled) | Manhattan (Unscaled) | Euclidean (Standard) | Manhattan (Standard) | Euclidean (Robust) | Manhattan (Robust) |
|:-----------|:---------------------|:---------------------|:---------------------|:---------------------|:-------------------|:-------------------|
| P1 vs P2   | 10000.0002     | 10002.0000     | 2.0178     | 2.8283     | 1.6667   | 2.3333   |
| P1 vs P3   | 5000.0004     | 5002.0000     | 1.4639     | 2.0265     | 1.2019   | 1.6667   |
| P2 vs P3   | 15000.0005     | 15004.0000     | 3.4330     | 4.8548     | 2.8284   | 4.0000   |


## Short reflection on scaling effects

### Subtask:
Provide a 3-4 sentence reflection answering: which scaler handled outliers better for the chosen features, whether scaling changed the relative closeness of points, and why this matters for algorithms that use distance.


## Short reflection on scaling effects

### Subtask:
Provide a 3-4 sentence reflection answering: which scaler handled outliers better for the chosen features, whether scaling changed the relative closeness of points, and why this matters for algorithms that use distance.

#### Reflection
For `Income_BDT`, `RobustScaler` handled the outlier (300,000) better by preserving its disproportionate magnitude (scaled to 12.0) compared to `MinMaxScaler` which compressed it to 1.0, thereby retaining its influence. Scaling dramatically altered the absolute distances between points, though in this specific subset, the *relative order* of closeness (P1-P3 being closest, P2-P3 farthest) was maintained for both Euclidean and Manhattan distances across unscaled, Standard Scaled, and Robust Scaled data. These changes are crucial because algorithms relying on distance, like K-Means or KNN, are sensitive to feature scales; without proper scaling, features with larger magnitudes can disproportionately dominate distance calculations, leading to biased or inaccurate model performance.

## Summary:

### Data Analysis Key Findings
*   DataFrames `C_Data_1` and `C_Data_2` were successfully combined into `combined_df`, resulting in a DataFrame of shape (5, 11) after dropping the 'ID' column.
*   A `ColumnTransformer` was effectively used to apply mixed preprocessing: one-hot encoding for nominal features (`Internet`, `City`), ordinal encoding for ordinal features (`Education`, `Satisfaction`, `Rating`), Min-Max scaling for `Age` and `GPA`, Standard scaling for `Hours_Study` and `Temp_C`, and Robust scaling for `Income_BDT` and `Transactions`.
*   The preprocessing transformed the data from an initial shape of (5, 11) to (5, 16), indicating the expansion due to one-hot encoding.
*   For the subset of 'Income_BDT' and 'Transactions', scaling drastically reduced the absolute Euclidean and Manhattan distances between points. For instance, the Euclidean distance between P1 and P2 decreased from `10000.0002` (unscaled) to `2.0178` (Standard Scaled) and `1.6667` (Robust Scaled).
*   While absolute distances changed significantly, the *relative order* of closeness between the three selected data points (P1, P2, P3) was preserved across unscaled, Standard Scaled, and Robust Scaled data.
*   `RobustScaler` was identified as handling the outlier in 'Income_BDT' (value 300,000) more effectively by preserving its disproportionate magnitude (scaled to 12.0), compared to `MinMaxScaler` which would compress it to 1.0.

### Insights or Next Steps
*   Feature scaling is crucial for distance-based machine learning algorithms (e.g., K-Means, KNN) as it prevents features with larger magnitudes from dominating distance calculations, leading to more accurate model performance.
*   The choice of scaler depends on data distribution and outlier presence; `RobustScaler` is particularly beneficial when dealing with datasets containing significant outliers.
